### PR TITLE
[DEVHAS-207] Detect ports in Component when running alizer

### PR DESCRIPTION
**Pull request by johnmcollier**
### What does this PR do?:
Updates our code in CDQs to retrieve the ports when we use alizer to detect Components. If ports are present for the detected component(s) they will be set in the Component stub. If no ports are detected, we will fall back on the ports present in the devfile.

### Which issue(s)/story(ies) does this PR fixes:
https://issues.redhat.com/browse/DEVHAS-207

### PR acceptance criteria:

- [ ] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

File: controllers/update.go [hunk 1]
```diff
@@ -327,6 +327,10 @@ func (r *ComponentDetectionQueryReconciler) updateComponentStub(req ctrl.Request
 			},
 		}
 
+		if len(componentPortsMap[context]) != 0 {
+			componentStub.TargetPort = componentPortsMap[context][0]
```
Comment (kim-tsao) [inline controllers/update.go:331]: does the first value always correspond to the first kubernetes component if set?

File: pkg/devfile/detect.go [hunk 2]
```diff
@@ -45,15 +45,16 @@ type AlizerClient struct {
 // Map 1 returns a context to the devfile bytes if present.
 // Map 2 returns a context to the matched devfileURL from the github repository. If no devfile was present, then a link to a matching devfile in the devfile registry will be used instead.
 // Map 3 returns a context to the dockerfile uri or a matched dockerfileURL from the devfile registry if no dockerfile is present in the context
-func search(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, error) {
+func search(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, map[string][]int, error) {
```
Comment (kim-tsao) [inline pkg/devfile/detect.go:48]: can we add a comment above the function to explain what Map 4 is?

File: pkg/devfile/detect.go [hunk 3]
```diff
@@ -147,11 +148,11 @@ func search(log logr.Logger, a Alizer, localpath string, devfileRegistryURL stri
 		err = &NoDevfileFound{Location: localpath}
 	}
 
-	return devfileMapFromRepo, devfilesURLMapFromRepo, dockerfileContextMapFromRepo, err
+	return devfileMapFromRepo, devfilesURLMapFromRepo, dockerfileContextMapFromRepo, componentPortsMapFromRepo, err
 }
 
 // AnalyzePath checks if a devfile or a dockerfile can be found in the localpath for the given context, this is a helper func used by the CDQ controller
-func AnalyzePath(a Alizer, localpath, context, devfileRegistryURL string, devfileMapFromRepo map[string][]byte, devfilesURLMapFromRepo, dockerfileContextMapFromRepo map[string]string, isDevfilePresent, isDockerfilePresent bool) error {
+func AnalyzePath(a Alizer, localpath, context, devfileRegistryURL string, devfileMapFromRepo map[string][]byte, devfilesURLMapFromRepo, dockerfileContextMapFromRepo map[string]string, componentPortsMapFromRepo map[string][]int, isDevfilePresent, isDockerfilePresent bool) error {
```
Comment (kim-tsao) [inline pkg/devfile/detect.go:155]: Can we add a comment to explain what the return values are?

File: pkg/devfile/detect.go [hunk 4]
```diff
@@ -269,20 +271,20 @@ func (a AlizerClient) DetectComponents(path string) ([]model.Component, error) {
 }
 
 // AnalyzeAndDetectDevfile analyzes and attempts to detect a devfile from the devfile registry for a given local path
-func AnalyzeAndDetectDevfile(a Alizer, path, devfileRegistryURL string) ([]byte, string, string, error) {
+func AnalyzeAndDetectDevfile(a Alizer, path, devfileRegistryURL string) ([]byte, string, string, []int, error) {
```
Comment (kim-tsao) [inline pkg/devfile/detect.go:274]: Can we add a comment to explain what the return values are?

File: pkg/devfile/devfile.go [hunk 5]
```diff
@@ -707,7 +707,7 @@ func DownloadDevfileAndDockerfile(url string) ([]byte, string, []byte) {
 // Map 1 returns a context to the devfile bytes if present.
 // Map 2 returns a context to the matched devfileURL from the devfile registry if no devfile is present in the context.
 // Map 3 returns a context to the dockerfile uri or a matched dockerfileURL from the devfile registry if no dockerfile is present in the context
-func ScanRepo(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, error) {
+func ScanRepo(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, map[string][]int, error) {
```
Comment (kim-tsao) [inline pkg/devfile/devfile.go:710]: Explain Map 4

File: controllers/update_test.go [hunk 6]
```diff
@@ -1272,9 +1272,9 @@ func TestUpdateComponentStub(t *testing.T) {
 			}
 			var err error
 			if tt.isNil {
-				err = r.updateComponentStub(ctrl.Request{}, nil, devfilesMap, nil, nil)
+				err = r.updateComponentStub(ctrl.Request{}, nil, devfilesMap, nil, nil, nil)
 			} else {
-				err = r.updateComponentStub(ctrl.Request{}, &componentDetectionQuery, devfilesMap, tt.devfilesURLMap, tt.dockerfileURLMap)
+				err = r.updateComponentStub(ctrl.Request{}, &componentDetectionQuery, devfilesMap, tt.devfilesURLMap, tt.dockerfileURLMap, nil)
```
Comment (kim-tsao) [inline controllers/update_test.go:1277]: can we pass in a not nil map of component ports here?

File: pkg/devfile/detect_test.go [hunk 7]
```diff
@@ -91,6 +92,15 @@ func TestAnalyzeAndDetectDevfile(t *testing.T) {
 			registryURL: DevfileStageRegistryEndpoint,
 			wantErr:     true,
 		},
+		{
+			name:                "Component detected successfully with Port(s) detected",
```
Comment (kim-tsao) [inline pkg/devfile/detect_test.go:96]: can we get coverage for a scenario where no ports are detected?

Comment (johnmcollier) [reply pkg/devfile/detect_test.go:96]: The other existing tests cover those scenarios, as a result of alizer's relatively fragile port detection

File: controllers/update.go (hunk 1 above)
Comment (johnmcollier) [reply controllers/update.go:331]: The ports here correspond to the ports returned by alizer, which were detected from the application source code - they should always be in order.

File: pkg/devfile/devfile.go (hunk 5 above)
Comment (johnmcollier) [reply pkg/devfile/devfile.go:711]: Updated.

File: pkg/devfile/detect.go (hunk 4 above)
Comment (johnmcollier) [reply pkg/devfile/detect.go:285]: Updated.

File: pkg/devfile/detect.go (hunk 3 above)
Comment (johnmcollier) [reply pkg/devfile/detect.go:161]: Updated.

File: pkg/devfile/detect.go (hunk 2 above)
Comment (johnmcollier) [reply pkg/devfile/detect.go:49]: Updated.

File: controllers/update_test.go (hunk 6 above)
Comment (johnmcollier) [reply controllers/update_test.go]: Fixed - good catch. Forgot to update this test when I was done testing. I added a test case to cover the scenario where a port is returned.

**Comment by johnmcollier** [comment]
@kim-tsao Thanks for the review - comments have been addressed.

**Pull request merged**